### ISSUE TITLE
Make WebTorrent more IPFS-friendly by better handling IPFS URLs used as webseeds

### ISSUE THREAD

**Issue by hollownights** [webtorrent/webtorrent#2434]
Make WebTorrent more IPFS-friendly by better handling IPFS URLs used as webseeds

When an URL for a file on the IPFS network is served through a [public IPFS gateway](https://docs.ipfs.tech/concepts/ipfs-gateway/) and used as source in a video element, the [IPFS companion browser extension](https://github.com/ipfs/ipfs-companion) can rewrite the request so that it points to the localhost, like this:

> https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR → http://localhost:8080/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR → http://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi.ipfs.localhost:8080

Unfortunately, the same isn’t true when WebTorrent process a public IPFS gateway URL used as a webseed for a torrent. In this case, the companion extension (apparently) doesn’t see the request to rewrite it and WebTorrent just tries to load the file through the public gateway, which is less than ideal, specially for large files.

With that said, I suggest that WebTorrent should better handle public IPFS gateway URLs as to work in conjunction with the IPFS companion extension, making possible for it to rewrite them as it already does for HTML elements. If that isn’t feasible right now or would take a long time to be implemented, then implement a workaround: when WebTorrent detects a public IPFS gateway URL used as webseed, automatically rewrite it to a path-like URL served by the localhost or, if it detects a native IPFS URL using the “ipfs:” protocol, rewrite it as a subdomain gateway URL. After rewriting it and making the request, wait for a 206 HTTP response. If it comes, just let the connection proceeds; if it doesn’t, then ditch the localhost URL and go back to the public gateway URL or, it was a native URL, just discard it. The [URLs would look like this](https://docs.ipfs.tech/how-to/address-ipfs-on-web/):

> https://gateway-host.tld/ipfs/{cid} → http://localhost:8080/ipfs/{cid}
> 
> ipfs://{cidv1} → http://{cidv1}.ipfs.localhost:8080

By making WebTorrent more IPFS-friendly some websites could more easily bridge the gap between a centralized web and a fully decentralized one: they could host all the static files on an IPFS node and publish them using their own public IPFS gateway and then, for audio and video files, they could also use WebTorrent, with the IPFS URLs serving as webseeds. 

To try out IPFS URLs as webseeds:

1) Install the [IPFS desktop](https://docs.ipfs.tech/install/ipfs-desktop/)
2) Install the [IPFS companion](https://docs.ipfs.tech/install/ipfs-companion/) on the browser you will use for testing; 
3) [Import a file](https://docs.ipfs.tech/basics/desktop-app/) (preferably a large file) to your local IPFS node; 
4) Create a torrent for that same file using instant.io or btorrent.xyz and get its magnet link;
5) Get a public IPFS gateway URL to that file using the [“Share link” option](https://docs.ipfs.tech/basics/desktop-app/#share-files);
6) Add that IPFS URL to the magnet link as a webseed by adding the parameter “ws=[URL]” to the magnet link (use “&” before/after it to add it to the magnet link);
7) Go and try to torrent it using the browser on which you installed the companion extension;
8) Now change the public IPFS gateway URL from the magnet link to the path-like localhost URL for the same file and try to torrent it again.

P.S.: I am not a developer, so I’m here as a user to propose such a feature, discuss it and test if it gets implemented.

**Comment by georgyo** [comment]
I like IPFS, but the question here is what is gained?

Specifically:

IPFS does not "store" files added to it in a persistent way. IE, in your example above the file would become unavailable as soon as you turned your computer off. You would need to pin it on a stable server for the file to persist, at which point that might as well be a webserver which would also consume far less compute and bandwidth resources for infrequent accesses. The amount of idle traffic the go-ipfs client creates just for it's DHT is staggering. 

The proposal is a one way fetch. If the file IS NOT findable on IPFS but is in the torrent swarm than this will not reseed IPFS causing requests to fail. It is also impossible to  recreate the IPFS DAG without the complete file AND knowing the parameters that were used to create the original DAG. 

If the file IS findable on IPFS and you have the desktop client installed, you will be downloading and seeding the file twice. Well parts of the file at the very least. Similarly, IPFS desktop makes it difficult to know when you STOP seeding. 

For a variety of reasons, native IPFS is slower than using a gateway. Which one of these has more trouble buffering and seeking? Though the second link will always be smoother on the second play trough after it is cached in your IPFS store; in the webtorrent case the caching is moot because you are also using webtorrent to download the file somewhere else.
`vlc https://bafybeih4ovrp5g65xtcpsf37uruwjzfituiumuvrusnzekehyixunql4dm.ipfs.dweb.link/manifest.m3u8`
vs
`vlc http://bafybeih4ovrp5g65xtcpsf37uruwjzfituiumuvrusnzekehyixunql4dm.ipfs.localhost:8080/manifest.m3u8`

There is already js-ipfs which allows downloading and seeding of IPFS content in the browser without the native client. If IPFS is your distribution platform of choice, commit to it.

**Comment by hollownights** [comment]
@georgyo Hi! Thanks for your answer! 

> IPFS does not "store" files added to it in a persistent way. IE, in your example above the file would become unavailable as soon as you turned your computer off.

Yes, I know, but this issue here isn't about IPFS, but really just about being able to get a file using Webtorrent/BitTorrent when a IPFS URL is used as webseed and being careful to not strain a public IPFS gateway like ipfs.io.

> You would need to pin it on a stable server for the file to persist, 

Yes, I know, and this is something I'm trying to cover with fully-fledged BitTorrent clients, like here: [After a torrent has finished downloading, make possible to (automatically) import its file/folder to a local IPFS node](https://github.com/frostwire/frostwire/issues/997).

> If the file IS NOT findable on IPFS but is in the torrent swarm than this will not reseed IPFS causing requests to fail.

Again, this issue here isn't about seeding the files on IPFS. See the issue above for something on that matter.

> If the file IS findable on IPFS and you have the desktop client installed, you will be downloading and seeding the file twice.

Except for the extra storage needed (which, like you said, is temporary), I see this as a plus.

> For a variety of reasons, native IPFS is slower than using a gateway.

Gotta say that I didn't experience this at all...

> Which one of these has more trouble buffering and seeking?

> > vlc https://bafybeih4ovrp5g65xtcpsf37uruwjzfituiumuvrusnzekehyixunql4dm.ipfs.dweb.link/manifest.m3u8

I couldn't load more than 10 seconds of video using this link, so I can't attest to what you said.

I'm under the impression that we've had really different experiences using IPFS, cause all of my experiences with it and with using an IPFS URL (localhost) as webseed to a torrent have been really smooth and fast. I've heard about its resources consumption just to keep running through the DHT, but that is above my pay-grade...

**Comment by ThaUnknown** [comment]
I've had these considerations before for WebTorrent, where people suggested backing WT's RTC seeders with stuff like IPFS, XDCC and Debrid, all of which I've turned down for a very simple reason: **_it is not webtorrent's place to extend those technologies_**

From what I understand the main issue stems from the "companion extension" which in itself doesn't handle the request correctly, and there's 0 reason for WT to fix that, you should file that issue on the companion extension's github

additionally adding custom webseeds in WT is actually REALLY easy, for example in just [200 LOC](https://github.com/ThaUnknown/wt-gdrive-webseed/blob/main/index.js) I created webseeds backed by a user's google drive, and a developer could easily add that custom functionality for IPFS too

admittedly WT's webseeds are very slow, like laughably so, this stems from the issue that WT doesn't correctly implement get-right webseeds as those are difficult to implement with the whole idea of "torrent streaming" that WT has going on, which in itself stems from the issue that browsers cannot handle HTTP request streams, instead the browser just pulls as much data, as fast as possible, ignoring if the code is actually consuming that data or not, leading to high RAM and CPU usage, so instead WT just pulls small chunks, but that introduces a lot of latency which makes the data download slow


TLDR: I don't think it's webtorrent that should handle IPFS, but some developer with a simple library to extend webtorrent with

**Comment by hollownights** [comment]
> I've had these considerations before for WebTorrent, where people suggested backing WT's RTC seeders with stuff like IPFS, XDCC and Debrid, all of which I've turned down for a very simple reason: **_it is not webtorrent's place to extend those technologies_**

@ThaUnknown That doesn't make any sense. BitTorrent, WebRTC and WebTorrent are not HTTP. Was adding support for HTTP webseeds extending the HTTP protocol?

> rom what I understand the main issue stems from the "companion extension" which in itself doesn't handle the request correctly, and there's 0 reason for WT to fix that, you should file that issue on the companion extension's github

I agree, the extension should be better and I will file an issue for that.

**Comment by ThaUnknown** [comment]
@hollownights unfortunately for your argument BitTorrent and WebTorrent are HTTP, as BitTorrent has a HTTP webseed spec, but not an ipfs one

**Comment by hollownights** [comment]
> as BitTorrent has a HTTP webseed spec

Common... If BitTorrent adds an IPFS spec then BitTorrent and WebTorrent will be IPFS?

**Comment by ThaUnknown** [comment]
Moving this conversation to https://github.com/webtorrent/webtorrent/issues/2467 as there are a lot of things wrong with webseeds so this unifies those conversations
